uk.co.compucorp.civicrm.hrcore/CRM/HRCore/Page/UserMenu.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+use CRM_HRCore_CMSData_PathsFactory as PathsFactory;
+
+class CRM_HRCore_Page_UserMenu extends CRM_Core_Page {
+
+  private $contactData;
+  private $cmsPaths;
+
+  public function run() {
+    $this->getContactData();
+    $this->instantiateCmsPaths();
+
+    $this->assign('username', $this->contactData['display_name']);
+    $this->assign('image', $this->getUserImagePath());
+    $this->assign('editLink', $this->cmsPaths->getEditAccountPath());
+    $this->assign('logoutLink', $this->cmsPaths->getLogoutPath());
+
+    return parent::run();
+  }
+
+  /**
+   * Gets the currently logged in contact's data, including the
+   * her user id in the CMS
+   *
+   * @return array
+   */
+  private function getContactData() {
+    $rawContactData = civicrm_api('Contact', 'getsingle', array(
+      'version' => 3,
+      'return' => array('id', 'display_name', 'image_URL'),
+      'id' => CRM_Core_Session::getLoggedInContactID(),
+      'api.User.getsingle' => array('contact_id' => "\$value.contact_id")
+    ));
+
+    $this->contactData = $this->normalizeContactDataAPIResponse($rawContactData);
+  }
+
+  /**
+   * Returns the path of the user's image, falling back to the CMS's default
+   * image if the user doesn't have one
+   *
+   * @return string
+   */
+  private function getUserImagePath() {
+    $defaultPath = $this->cmsPaths->getDefaultImagePath();
+
+    if (isset($this->contactData['image_URL']) && !empty($this->contactData['image_URL'])) {
+      return $this->contactData['image_URL'];
+    } else {
+      return $defaultPath;
+    }
+  }
+
+  /**
+   * Instantiates the paths class of the current CMS
+   */
+  private function instantiateCmsPaths() {
+    $cmsName = CRM_Core_Config::singleton()->userFramework;
+
+    $this->cmsPaths = PathsFactory::create($cmsName, $this->contactData);
```

Comment:
This method uses `$this->contactData` which is only set when you call `getContactData()`, so you have a hidden dependency here. Instead of accessing the property directly, use the method. If you're concerned about making an API call everytime you call if, you can implement it like this:

```php
public function getContactData() {
    if(!empty($this->contactData)) {
       $this->contactData = getDataFromAPI();
    }

    return $this->contactData;
}
```